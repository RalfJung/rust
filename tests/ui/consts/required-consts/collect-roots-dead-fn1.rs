//@revisions: noopt opt
//@ build-fail
//@[noopt] compile-flags: -Copt-level=0
//@[opt] compile-flags: -O

//! This used to fail in optimized builds but pass in unoptimized builds. The reason is that in
//! optimized builds, `f` gets marked as cross-crate-inlineable, so the functions it calls become
//! reachable, and therefore `g` becomes a collection root. But in unoptimized builds, `g` is no
//! root, and the call to `g` disappears in an early `SimplifyCfg` before "mentioned items" are
//! gathered, so we never reach `g`.
#![crate_type = "lib"]

struct Fail<T>(T);
impl<T> Fail<T> {
    const C: () = panic!(); //~ERROR: evaluation of `Fail::<i32>::C` failed
}

pub fn f() {
    loop {}; g()
}

#[inline(never)]
fn g() {
    h::<i32>()
}

// Make sure we only use the faulty const in a generic function, or
// else it gets evaluated by some MIR pass.
#[inline(never)]
fn h<T>() {
    Fail::<T>::C;
}
