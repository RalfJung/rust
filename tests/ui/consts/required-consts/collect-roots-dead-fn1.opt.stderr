error[E0080]: evaluation of `Fail::<i32>::C` failed
  --> $DIR/collect-roots-dead-fn1.rs:15:19
   |
LL |     const C: () = panic!();
   |                   ^^^^^^^^ the evaluated program panicked at 'explicit panic', $DIR/collect-roots-dead-fn1.rs:15:19
   |
   = note: this error originates in the macro `$crate::panic::panic_2015` which comes from the expansion of the macro `panic` (in Nightly builds, run with -Z macro-backtrace for more info)

note: erroneous constant encountered
  --> $DIR/collect-roots-dead-fn1.rs:31:5
   |
LL |     Fail::<T>::C;
   |     ^^^^^^^^^^^^

note: the above error was encountered while instantiating `fn h::<i32>`
  --> $DIR/collect-roots-dead-fn1.rs:24:5
   |
LL |     h::<i32>()
   |     ^^^^^^^^^^

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0080`.
